design doc addition (#386)
microsoft/vscode-python-environments: docs/design.md

# Python Environments API

> The following helps describe the technical design of the Python Environments Extension for use mainly by contributing extension authors. This extension is the entry point which will surface the entire environment management experience for Python in VS Code and its modularity should allow for individual tools / libraries to implement these methods for their tool making them usable in VS Code.

Reviewer: ```suggestion
The following guide describes the technical design of the Python Environments Extension. Its intended use is mainly for contributing extension authors. This extension is the entry point which will surface the entire environment management experience for Python in VS Code and its modularity should allow for individual tools / libraries to implement these methods for their tool making them usable in VS Code.
```

**Table of Contents**
- [Python Environments API](#python-environments-api)
  - [Functionalities of Extension \& its API](#functionalities-of-extension--its-api)
  - [API Surface](#api-surface)
  - [Objects in Extension API](#objects-in-extension-api)
  - [Relationships between API Components](#relationships-between-api-components)
    - [Workspace and Environments](#workspace-and-environments)
    - [EnvironmentManagers and PackageManagers](#environmentmanagers-and-packagemanagers)
    - [Packages and Related Objects](#packages-and-related-objects)
    - [PythonProject and PythonProjectCreator](#pythonproject-and-pythonprojectcreator)
  - [Connection to UI](#connection-to-ui)
    - [Python Side Pane](#python-side-pane)
    - [Copilot Tooling](#copilot-tooling)
    - [Command Palette](#command-palette)
    - [Contributions to File Explorer](#contributions-to-file-explorer)

## Functionalities of Extension & its API

> The python environment extension has four main functionalities: 

Reviewer: ```suggestion
The Python Environments extension has four main functionalities: 
```

- `EnvironmentManager`: an object which manages `PythonEnvironment`
    - registered by an extension in form with id `<publisher-id>.<extension-id>:<manager-name>`
    - acts on an environment (get, set, etc)  `PythonEnvironment`
    - one `EnvironmentManager` can create many `PythonEnvironment` objects
- `PackageManagementOptions`: actions relating to, and specifying packages
- `PackageManager`: an object which manages packages
    - registered by an extension in form with id `<publisher-id>.<extension-id>:<manager-name>`
    - performs actions for packages using `PackageManagementOptions`

## Relationships between API Components
Generally in the extension, the PackageManagers and EnvironmentManagers handle the entry-point for contributing actions in the area of getting and setting environments as well as accessing activation / execution events. Projects can be created via PythonProjectCreators but ongoing management of projects in a workspace are handled through the envs ext API.

### Workspace and Environments
The relationship between **workspaces** and **environments** in the extension can be any of the following configurations:

> | workspace | environment | description | example |

Reviewer: ```suggestion
| Workspace | Environment | Description | Example |
```
